### PR TITLE
ratelimit: per descriptor hits addend support and prefer uint64

### PR DESCRIPTION
**Pull request by wbpcode**

### CODE REVIEW

**Review by mathetake — COMMENTED** [review]
high level question: if the number zero is given, it should work like "check the budget is not negative, and if not requests are allowed"? I am asking because if this would obviate the issue we had in https://github.com/envoyproxy/ratelimit/pull/729

**Comment by wbpcode** [comment]
> high level question: if the number zero is given, it should work like "check the budget is not negative, and if not requests are allowed"? I am asking because if this would obviate the issue we had in #729

The counter in redis is a number that counts all used tokens. So, we only need to check if the counter is larger than the threshold or not.

**Comment by mathetake** [comment]
cool, so in other words, sending hits_addend=0 can be used to reject requests if the counter in redis is negative or zero ya?

**Comment by wbpcode** [comment]
> cool, so in other words, sending hits_addend=0 can be used to reject requests if the counter in redis is negative or zero ya?

sorry, I may didn't make it clear. The counter will never be negative in the redis. It is a counter the will increases forever in the time window.

But the conclusion is right. hits_addend zero could also could be used to reject a request.

**Comment by mathetake** [comment]
great, thank you for clarifying!

**Review by mathetake — APPROVED** [review]

File: src/redis/fixed_cache_impl.go [hunk 1]
```diff
@@ -32,119 +32,131 @@ type fixedRateLimitCacheImpl struct {
...
 	*pipeline = client.PipeAppend(*pipeline, result, "INCRBY", key, hitsAddend)
 	*pipeline = client.PipeAppend(*pipeline, nil, "EXPIRE", key, expirationSeconds)
 }
 
-func pipelineAppendtoGet(client Client, pipeline *Pipeline, key string, result *uint32) {
+func pipelineAppendtoGet(client Client, pipeline *Pipeline, key string, result *uint64) {
 	*pipeline = client.PipeAppend(*pipeline, result, "GET", key)
 }
 
+func (this *fixedRateLimitCacheImpl) getHitsAddend(hitsAddend uint64, isCacheKeyOverlimit, isCacheKeyNearlimit,
```
Comment (mathetake) [inline src/redis/fixed_cache_impl.go:44]: do you want to add unit tests for this function?

Comment (wbpcode) [reply src/redis/fixed_cache_impl.go:44]: Of cource. I will add some unit tests for the whole feature.

**Comment by wbpcode** [comment]
cc @psbrar99

**Comment by mathetake** [comment]
also @zirain if you have any comments

File: src/utils/utilities.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ func CalculateReset(unit *pb.RateLimitResponse_RateLimit_Unit, timeSource TimeSo
 	return &durationpb.Duration{Seconds: sec - now%sec}
 }
 
-func Max(a uint32, b uint32) uint32 {
+func Max(a uint64, b uint64) uint64 {
```
Comment (zirain) [inline src/utils/utilities.go:44]: not a must, but is it find to remove this as golang support generic?

**Review by zirain — APPROVED** [review]
LGTM, only a small nit.

File: src/utils/utilities.go (hunk 2 above)
Comment (zirain) [reply src/utils/utilities.go:44]:
```
// The max built-in function returns the largest value of a fixed number of
// arguments of [cmp.Ordered] types. There must be at least one argument.
// If T is a floating-point type and any of the arguments are NaNs,
// max will return NaN.
func max[T cmp.Ordered](x T, y ...T) T
```

**Comment by psbrar99** [comment]
LGTM, thanks @wbpcode .

**Review by psbrar99 — APPROVED** [review]

**Comment by mathetake** [comment]
fantastic, thank you for your hard work! @wbpcode 👯

**Pull request merged**